### PR TITLE
feat: upgrade of the client library

### PR DESCRIPTION
**Pull request by toddb**
@wh1337 I am working on getting a working version of this client into my own product and here is a first step to see if you are interested in this. Let me know—I've generally tried to follow your style but not always. Note: at this stage, I have attempted to sync the readme.

 - https://github.com/novuhq/novu-dotnet/issues/34 - injectable
 - https://github.com/novuhq/novu-dotnet/issues/45
 - added DI for client for tests and extensions for production applications
 - reworked the tests between integration and micro tests
 - added workflows (and steps) can now be added reasonably stably
 - upgraded client to allow for either HttpClient or RestService

### CODE REVIEW

**Comment by toddb** [comment]
@wh1337 Just added the Layout #21

**Comment by toddb** [comment]
@wh1337 Just added the Integration client https://github.com/novuhq/novu-dotnet/issues/20

It's a funny lifecycle so the testing is a bit ad hoc for me.

**Comment by toddb** [comment]
@wh1337 Added subscriber notifications and a test demonstrating a full lifecycle

**Comment by toddb** [comment]
I have now put in Topic triggers to go alongside Event triggers—there are end to end tests on that too. I will be beefing up scenarios because this is what I have been aiming for to integration into my product.

**Comment by toddb** [comment]
@wh1337 I am getting to the point that I need to incorporate this functionality into my product.  How are you getting on? 

I would prefer to avoid releasing a separate nuget package.

**Comment by unicodeveloper** [comment]
@toddb this will be resolved today. @wh1337 please what's the delay?

File: src/Novu/Interfaces/ISubscriberClient.cs [hunk 1]
```diff
@@ -1,66 +1,91 @@
...
 
     /// <summary>
-    /// Get a single Subscriber
+    ///     Get a single Subscriber
     /// </summary>
-    /// <param name="id"><see cref="String"/> Subscriber ID</param>
+    /// <param name="id"><see cref="String" /> Subscriber ID</param>
     /// <returns></returns>
     [Get("/subscribers/{id}")]
     Task<SubscriberDto> GetSubscriber(string id);
```
Comment (joaonlforte) [inline src/Novu/Interfaces/ISubscriberClient.cs:29]: Returning a single subscriber is now wrapped in a `data` prop, so this is broken atm 

curl -X GET \
 -H "Authorization: ApiKey <api_key>" \
https://api.novu.co/v1/subscribers/xyz | jq

```js
{
  "data": { <== this has changed
     "subscriberId": "xyz"
    ...
    ...
  }
}
```

Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: @joaonlforte I'm struggling to reproduce the error in the tests on the [fork](https://github.com/picmi/novu-dotnet)

I have 0.17.1 running (this [docker script](https://github.com/novuhq/novu/blob/53d2acde2180dc3c8b5da1a620b0d5c3fd54a306/docker/local/deployment/docker-compose.yml#L26))

[This test](https://github.com/picmi/novu-dotnet/blob/2b29f957b91d5a64089475850a2697b3116fc9f6/src/Novu.Tests/IntegrationTests/SubscriberTests.cs#L30) passes

I know that the original pull request was against 0.16 because that was the latest at that point. I then had to upgrade I think about 2-3 days later to get some of the other commits to pass. 

Personally, the test project needs to know what version it is running against.

Thanks in advance to help me work out what is wrong (or what I am doing wrong).

**Comment by toddb** [comment]
@wh1337 All good. I'm sorry that the work is far from perfect but it gets some more key patterns in play (particularly the polymorphic JSON mapping). At first, the pull request was to see if you are interested and then it took a life of its own and I needed much more for provisioning up the system. Personally, I think some tweaks to patterns particularly around naming would make it a more consumable interface. But this is a wider discussion—and part of the work I am doing to write a provisioning agent as the basis later on for a terraform provider.

Much depends on your direction you want :-) Cheers. (you call always DM me on discord as I know this isn't the right place for this discussion. Again sorry.)

**Comment by toddb** [comment]
@wh1337 I will need to stand up a fresh environment to check but it looks like not enough filtering going on with you environment. Personally, integration test should be able to cater for existing data even if they setup themselves up

I honestly thought I have enough setup and teardown that disregarded the existing environment (except for layouts and integrations as they have global lifecylces). 

Re: 3 errors

- two look to be setup data tweaking (elements don't match)
- other one (execution details) looks like more complicated but I suspect again data setup

Note: you can create a `appsettings.[USER_PROFILE].json` so that you don't pollute the Integration version for the build pipeline (then add an ignore in the .gitignore—it's not pretty).

**Comment by toddb** [comment]
@wh1337 If you look inside the configuration setup, you can still use environment variables that you have—I tried not to break your work unnecessarily—alternatively, Integration settings can have it checked in (but it seems injecting is good). What I have done here is implement a more idiomatic configurations, particularly if you have aspnet/core pipelines.

Re: versioning. The versioning is more nuanced that "v1". My belief is that it is actually the product version 0.17.x (given I have already seen one upgrade go on in the last week). I have NOT looked at tests that detect this version, nor I suspect looking at being stable across multiple versions.

File: src/Novu/Interfaces/ISubscriberClient.cs (hunk 1 above)
Comment (joaonlforte) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: Hi @toddb 👋 

I can't tell exactly the version, since I'm using the SaaS version of Novu (cloud), and i'm using [this](https://www.nuget.org/packages/Novu) package v0.2.2

The error i'm getting upon deserialising the response is 

```
Refit.ApiException: An error occured deserializing the response.
 ---> Newtonsoft.Json.JsonSerializationException: Cannot deserialize the current JSON object (e.g. {"name":"value"}) into type 'System.Collections.Generic.List`1[Novu.DTO.AdditionalDataDto]' because the type requires a JSON array (e.g. [1,2,3]) to deserialize correctly.
To fix this error either change the JSON to a JSON array (e.g. [1,2,3]) or change the deserialized type so that it is a normal .NET type (e.g. not a primitive type like integer, not a collection type like an array or List<T>) that can be deserialized from a JSON object. JsonObjectAttribute can also be added to the type to force it to deserialize from a JSON object.
Path 'data._id', line 1, position 15.
```

but the part i find odd is this one `Path 'data._id', line 1, position 15.`. It seems that the entire response of the `GetSubscriber` is now wrapped in a `data` property (similarly what is happening with the `CreateTopic` endpoint where a `TopicCreateResponseDto` is returned )

Screenshot 2023-08-22 at 10 35 02

In summary seems that the response contract has changed for the Subscriber endpoints and now the actual response object is wrapped in a `Data` prop

Screenshot 2023-08-22 at 10 38 31
Screenshot 2023-08-22 at 10 38 50

So instead of returning a `SubscriberDto` should be returning it in an envelope, something like

```csharp 
class Foo {
    [JsonProperty("data")]
    public SubscriberDto Data { get; set; }
}
```
```csharp

    [Get("/subscribers/{id}")]
    Task<Foo> GetSubscriber(string id);

    [Post("/subscribers")]
    Task<Foo> CreateSubscriber([Body] CreateSubscriberDto dto);

```

Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: I really suspect you are running against 0.16.x and the client code I am writing against is 0.17.x. I can imagine the frustration. I have had a quick look and I can see no versioning of the API via the API responses. 

You have introduced the class `Foo` but what you are looking for is `NovuResponse<T>` or `PaginatedResponseDto<T>`. However, where confusion lies IMHO is that in some cases neuther are used and the `Data<T>` is inline (eg `TriggerBulkResponseDto` or `SubscribersDto`). 

I hope there is some help here.

ps. when I had this problem, I had to get out the source to check what as going on. Sorry.

**Comment by toddb** [comment]
@wh1337 I have pushed some changes that make the tests less flaky. Running both the acceptance and the integration tests in parallel in all likelihood lead to tears. You'll see that I have written them with Generators to create uniqueness/separation but in practice there are still collisions. My money is that if you run each individually (or sequentially) they will pass. 

The round round of assertions need to be less specific assertions and more a matching style. I think the tests need to (a) confirm serialisation against a version of the API (integration tests) and then (b) detect lifecycle collisions (acceptance tests). 

I will only be able to "fix" against the cloud-based solutions with credentials.

Cheers.

File: src/Novu/Interfaces/ISubscriberClient.cs (hunk 1 above)
Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: @joaonlforte So, I have found where the API is versioned. It is on the `health-check` resource. Looking at the source, it is not an external API resource (so you can't use the ApiKey token) and need to hijack your JWT. I did that by copying a request and then changing it. I don't get why this isn't at the root of the API.

```
  ~ curl 'http://localhost:3000/v1/health-check' \
  -H 'Accept: application/json, text/plain, */*' \
   -H 'Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2NGM5YTc1ZmUxOGQ3ZjJjYjk3ODcxNjMiLCJmaXJzdE5hbWUiOiJ0IiwibGFzdE5hbWUiOiJub3Z1IiwiZW1haWwiOiJ0K25vdnUtbG9jYWwuZGV2QHBpY21pLmlvIiwib3JnYW5pemF0aW9uSWQiOiI2NGM5YTc3NWUxOGQ3ZjJjYjk3ODcxNzAiLCJyb2xlcyI6WyJhZG1pbiJdLCJlbnZpcm9ubWVudElkIjoiNjRjOWE3NzZlMThkN2YyY2I5Nzg3MTc2IiwiaWF0IjoxNjkyNTc0NjU1LCJleHAiOjE2OTUxNjY2NTUsImlzcyI6Im5vdnVfYXBpIn0.9-iVXXsWKfR7DaBojrhlMtRjMBjXSAJGiZXt0GIxZYU' \
  --compressed
{"status":"error","info":{"db":{"status":"up"},"apiVersion":{"version":"0.17.1","status":"up"},"triggerQueue":{"status":"up"}},"error":{"inMemory":{"status":"down"}},"details":{"db":{"status":"up"},"apiVersion":{"version":"0.17.1","status":"up"},"triggerQueue":{"status":"up"},"inMemory":{"status":"down"}}}%  
```

Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: So the answer in production is: 0.18.0

```
~ curl 'https://api.novu.co/v1/health-check' \
  -H 'authority: api.novu.co' \
  -H 'accept: application/json, text/plain, */*' \
  -H 'accept-language: en-GB,en-US;q=0.9,en;q=0.8,mi;q=0.7' \
  -H 'authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6I ...-bfSmesKN2w' \
   --compressed
   
{"data":{"status":"ok","info":{"db":{"status":"up"},"inMemory":{"status":"up"},"apiVersion":{"version":"0.18.0","status":"up"},"triggerQueue":{"status":"up"},"cacheService":{"status":"up"}},"error":{},"details":{"db":{"status":"up"},"inMemory":{"status":"up"},"apiVersion":{"version":"0.18.0","status":"up"},"triggerQueue":{"status":"up"},"cacheService":{"status":"up"}}}}%                 ➜  ~ 
```

Comment (joaonlforte) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: > I really suspect you are running against 0.16.x and the client code I am writing against is 0.17.x. I can imagine the frustration. I have had a quick look and I can see no versioning of the API via the API responses.
> 
> You have introduced the class `Foo` but what you are looking for is `NovuResponse<T>` or `PaginatedResponseDto<T>`. However, where confusion lies IMHO is that in some cases neuther are used and the `Data<T>` is inline (eg `TriggerBulkResponseDto` or `SubscribersDto`).
> 
> I hope there is some help here.
> 
> ps. when I had this problem, I had to get out the source to check what as going on. Sorry.

Thanks @toddb 

that might be the issue but i don't own the client, so i can't change the contract, I was hoping a fix would be released. Aren't you the maintainer of the package?

Thanks in advance ;)

Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs:29]: Nope. I'm not. I've just made a pull request so that it works for me! But you are right that it is broken for 0.18. I will put through a fix in the next 24 hours (after I have a sleep).

**Comment by toddb** [comment]
> Running it against `https://api.novu.co/v1` i'm getting 23 errors

I have confirmed that these are because the API is v0.18.x. The API has been "upgraded".

File: src/Novu/Interfaces/ISubscriberClient.cs (hunk 1 above)
Comment (toddb) [reply src/Novu/Interfaces/ISubscriberClient.cs]: @joaonlforte Thanks—I upgraded to 0.18 (which coincidentally is not updated in the docker-compose[ deployment script](https://github.com/novuhq/novu/blob/next/docker/local/deployment/docker-compose.yml)). 

This library is broken against 0.18.0.

@wh1337 It seems that this library needs to be able create the test environment locally. That means maintaining copy of the docker-compose doesn't it?

**Comment by toddb** [comment]
@joaonlforte @wh1337 Fix for new interface is now in the [commit aa0b95b6f](https://github.com/novuhq/novu-dotnet/pull/46/commits/aa0b95b6f43c0bc12deeeac7badd69570290f621).

Looking at the source, I can't actually see where the change comes from. 

Note: this does suggest the need for a refactor that I have been quiet on (but will be a breaking a change (ie code upgrade) for all users of the library). Ie remove buried 'data' and use `NovuResponse<T>` and `PaginatedResponse<T>` more consistently throughout. While I am on a rant, I would drop the Dto convention and think of the interface more in the repository pattern ...

**Comment by joaonlforte** [comment]
> @joaonlforte @wh1337 Fix for new interface is now in the [commit aa0b95b6f](https://github.com/novuhq/novu-dotnet/pull/46/commits/aa0b95b6f43c0bc12deeeac7badd69570290f621).
> 
> Looking at the source, I can't actually see where the change comes from.
> 
> Note: this does suggest the need for a refactor that I have been quiet on (but will be a breaking a change (ie code upgrade) for all users of the library). Ie remove buried 'data' and use `NovuResponse<T>` and `PaginatedResponse<T>` more consistently throughout. While I am on a rant, I would drop the Dto convention and think of the interface more in the repository pattern ...

Thanks @toddb for the update on this. Do you have an ETA when would a new version this fix will be available? 

Thanks for your support 🙏

**Comment by toddb** [comment]
> @toddb
> 
> > Note: this does suggest the need for a refactor that I have been quiet on (but will be a breaking a change (ie code upgrade) for all users of the library). Ie remove buried 'data' and use NovuResponse and PaginatedResponse more consistently throughout. While I am on a rant, I would drop the Dto convention and think of the interface more in the repository pattern ...
> 
> I agree with this. This is my first true SDK for c# so I'm all for other perspectives. I'll make a note of the breaking change for the SDK version so when we make the release notes the end users know about it.
> 
> Did you want to add the DTO conversion to this PR or would you like to seperate that out?

@wh1337 Thanks for the clear go ahead! I will do the changes now. It seems people are waiting but let them have a better change in one.

**Comment by toddb** [comment]
@wh1337 sorry to spam, the plan of attack (should help with migration documentation) 

* https://github.com/novuhq/novu-dotnet/issues/49

**Comment by toddb** [comment]
@joaonlforte @wh1337 Refactor based on #49 is now in the pull.

@wh1337 the documentation will need updating (particularly the readme)

**Comment by toddb** [comment]
@wh1337 given you are talking about this as an SDK then I have boldly included the 'Sync' functionality (#48) for this release. It can be split off easily if this is not wanted—let me know.

Note: The acceptance test for this functionality is "hidden" behind only being able to be run in debug mode.

**Comment by toddb** [comment]
> 



> Thanks @toddb for the update on this. Do you have an ETA when would a new version with this fix be available?

I have made all the changes I think are needed and it is over to @wh1337 to do the merge/release. I don't have rights.

**Comment by toddb** [comment]
> @toddb fantastic work with this. Give me a little to run through and ensure the tests are passing. Before we release, we will need to write up some extensive change notes to ensure downstream users are fully aware of what is coming.

There are people waiting and I would release early and follow fast with documentation. Get people with problems to come to Novu Discord. The naming conventions should be straightforward enough once you get a hang of it. But mainly it is now consistent! (jury out on good :-))

**Comment by toddb** [comment]
> With that, can you tell me how you go about your testing? The more details the better as I will use that to work on test automation via Github Actions.

Discord? (todd7201) Not sure what the gap is.  Testing Pyramid and Testing Quadrants models for theory, also the GOOS style.

* Strategy is micro. integration, acceptance testing separation (normally, I would separate by projects and in this case we could or use Traits). 
* Each test has separation for parallelism—in the integration tests you can see I have gone to lengths to use MotherObject-type patterns to manage lifecycle setup/teardowns (it is in the base class which IMHO is a bit messy but never bothered to find a better option)
* Mother objects go at length to generate readable uniqueness (hence using ParkSquare generators)
* Acceptance tests are the pattern that it is not testing one thing but rather a sequence

Most importantly, I use the tests as the first client to maintain the usability of the clients—understandability and setting injection. Hence why I struggled with the first version as I could rarely guess the classes I needed, etc.

Something of use in that?

**Comment by toddb** [comment]
The test will fail without the correct injection into Actions. 

You can inject as environment variables from the command line run.

```
name: Test
        working-directory: ./src/Novu.Tests
        run: dotnet test  -p:Novu__Url=https://api.novu.co/v1 -p:Novu__ApiKey=${{ secrets.NOVU_API_KEY }}
 ```

**Comment by joaonlforte** [comment]
@wh1337 @toddb it would be amazing to get this out the door as soon as possible. Given the latest breaking changes of the api, we now have notifications broken and blocked, we want to go live and this is a major blocker atm.

Can you guys provide a realistic ETA for this version get released? We need to manage expectations internally on this.

Thanks and keep up the good work ;)

**Comment by toddb** [comment]
> Given the latest breaking changes of the api, we now have notifications broken and blocked, we want to go live and this is a major blocker atm.

@joaonlforte of course you are blocked in production—didn't even occur to me. Fixing that shouldn't have been part of this pull request. It should be raised as an issue against Novu (0.18.0) for release 0.2.2, patched and released immediately. Apologies for my part in this because that an outage for you. Now an issue #50.

Question then, what versioning scheme would you like to see on this project that helps manage versioning against the Novu API? This will happen again and a clean, clear process for reporting, diagnostics and then resolution. Can you see anything that the core Novu team need to do too? (eg the 0.18.0 release wasnt' even notified, let alone a breaking interface or a convention that suggests it (eg odd/even)). Now an issue #51. Please comment there as it will get lost here.

Finally, @joaonlforte the changes coming through are a rewrite (you will need to factor this in). Is this going to be a problem? If so, @wh1337 should patch the code and re-release FIRST. Otherwise, you should get a copy of the fork and factor your code on a locally compiled version before it is released and get it ready. Otherwise, you could have more downtime. If you wanted help with the migration, I am happy to help to speed things up—the interface conventions are documented  here at #49 but there is no migration guide as yet. I am on the novu discord channel (todd7201).

**Comment by joaonlforte** [comment]
> @joaonlforte @toddb @unicodeveloper
> 
> This has been merged to main. Give @toddb and myself a little time to get release documenation together and updated, and I will publish a new version on Nuget.
> 
> **Important:** This new release will introduce breaking changes.

Hi @wh1337 

I've updated the package and did the necessary changes. All seems to be working fine now with 0.18.0.

Thanks ;)
